network/p2p/libp2pNode.go

```diff
@@ -528,32 +572,35 @@ func bootstrapLibP2PHost(ctx context.Context,
 	// gather all the options for the libp2p node
 	options := []config.Option{
 		libp2p.ListenAddrs(sourceMultiAddr), // set the listen address
-		libp2p.Identity(key),                // pass in the networking key
-		libp2p.ConnectionManager(conMgr),    // set the connection manager
+		libp2p.Identity(libp2pKey),          // pass in the networking key
 		transport,                           // set the protocol
-		libp2p.Ping(true),                   // enable ping
+		libp2p.Ping(pingEnabled),            // enable ping
 	}
 
-	// if allowlisting is enabled, create a connection gator with allowListAddrs
-	if allowList {
-		// create a connection gater
-		connGater = newConnGater(logger)
+	return options, nil
+}
 
-		// provide the connection gater as an option to libp2p
-		options = append(options, libp2p.ConnectionGater(connGater))
-	}
+// WithDefaultPubSub returns a NodeOption that initializes a GossipSub object for the node libp2p host with the given
+// options
+func WithDefaultPubSub(psOption ...pubsub.Option) NodeOption {
+	return func(node *Node) error {
 
-	// create the libp2p host
-	libP2PHost, err := libp2p.New(ctx, options...)
-	if err != nil {
-		return nil, nil, nil, fmt.Errorf("could not create libp2p host: %w", err)
+		// Creating a new PubSub instance of the type GossipSub with psOption
+		pubSub, err := pubsub.NewGossipSub(node.ctx, node.host, psOption...)
+		if err != nil {
+			return fmt.Errorf("could not create libp2p pubsub: %w", err)
+		}
+		node.pubSub = pubSub
+		return nil
 	}
+}
 
-	// Creating a new PubSub instance of the type GossipSub with psOption
-	ps, err := pubsub.NewGossipSub(ctx, libP2PHost, psOption...)
-	if err != nil {
-		return nil, nil, nil, fmt.Errorf("could not create libp2p pubsub: %w", err)
+// WithDefaultPingService returns a NodeOption that initializes the PingService object for the node
+func WithDefaultPingService(rootBlockID string, pingInfoProvider PingInfoProvider) NodeOption {
+	return func(node *Node) error {
+		pingLibP2PProtocolID := generatePingProtcolID(rootBlockID)
+		pingService := NewPingService(node.host, pingLibP2PProtocolID, pingInfoProvider, node.logger)
+		node.pingService = pingService
+		return nil
 	}
```

Comment:
For the sake of my understanding, could you explain for each of these `WithDefault...` functions, what are the fields that could actually potentially be different from the default?

Reply:
I'm wondering if instead of having `WithPingService`, `WithPubSub`, `WithConnectionManager` etc we could just have a single `WithHost`?

Reply:
For example, what would a non-default `pingService` or `pubsub` look like?

Reply:
a non default pubsub would be using one of the other pubsub techniques that libp2p provides (floodsub or randomsub) instead of pubsub. A non default example of a pingService would be a custom ping service being passed in.